### ISSUE TITLE
stream_unzip.UnexpectedSignatureError

### ISSUE THREAD

**Issue by vitkovay** [uktrade/stream-unzip#33]
When decompressing a zip file this error is raised:

<code>stream_unzip.UnexpectedSignatureError: b'\xb66\x97\x89'</code>

Part of my code:
<code>

    def read_in_chunks(infile, chunk_size=UNZIP_CHUNK_SIZE):
        while True:
            chunk = infile.read(chunk_size)
            if chunk:
                yield chunk
            else:
                return<br>
    with open(path_to_zip_file, 'rb') as file_object:
        for file_name, file_size, file_chunks in stream_unzip(read_in_chunks(file_object), chunk_size=UNZIP_CHUNK_SIZE):
            print(file_name)
            with open(file_name, 'ab') as file_append:
                for chunk in file_chunks:
                    file_append.write(chunk)
</code> 

Python 3.9
stream-unzip 0.0.70

I can unzip my file with WinRAR and [zipfile](https://docs.python.org/3.9/library/zipfile.html), but can't with stream-unzip. 

Any idea how fix it?

**Comment by vitkovay** [comment]
My zip file was created ZipOutputStream from [JDK](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/zip/ZipOutputStream.html).

**Comment by vitkovay** [comment]
My zip file was created ZipOutputStream from [JDK](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/zip/ZipOutputStream.html).

**Comment by vitkovay** [comment]
My zip file was created ZipOutputStream from [JDK](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/zip/ZipOutputStream.html).

**Comment by michalc** [comment]
Hi @vitkovay,

Thanks for the report. However, I've tried creating a zip file using ZipOutputStream, but for me it works fine.

**Zip.java**
```java
import java.io.FileOutputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.BufferedOutputStream;
import java.util.zip.ZipOutputStream;
import java.util.zip.ZipEntry;

class Zip {
    public static void main(String[] args) throws FileNotFoundException, IOException {
        FileOutputStream fos = new FileOutputStream("file.zip");
        BufferedOutputStream bos = new BufferedOutputStream(fos);
        ZipOutputStream zos = new ZipOutputStream(bos);

        try {
            zos.putNextEntry(new ZipEntry("file-a.txt"));
            zos.write("contents-a".getBytes());
            zos.closeEntry();
            zos.putNextEntry(new ZipEntry("file-b.txt"));
            zos.write("contents-b".getBytes());
            zos.closeEntry();
        }
        finally {
            zos.close();
        }
    }
}
```

**j.py**
```python
from stream_unzip import stream_unzip

def get_compressed_chunks():
    with open('file.zip', 'rb') as f:
        while c := f.read(1024):
            yield c

for name, size, uncompressed_chunks in stream_unzip(get_compressed_chunks()):
    for chunk in uncompressed_chunks:
        print(chunk)
```
Then I run:
```bash
javac Zip.java
java Zip
python j.py
```
and I get the expected output of:
```
b'contents-a'
b'contents-b'
```

Are you able to post a similar reproducible example that I can run to get the same error?

**Comment by michalc** [comment]
Ah thanks - will investigate

**Comment by michalc** [comment]
I think I can reproduce it, using this to make the file:
```java
import java.io.FileOutputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.BufferedOutputStream;
import java.util.zip.ZipOutputStream;
import java.util.zip.ZipEntry;

class Zip {
    public static void main(String[] args) throws FileNotFoundException, IOException {
        FileOutputStream fos = new FileOutputStream("file.zip");
        BufferedOutputStream bos = new BufferedOutputStream(fos);
        ZipOutputStream zos = new ZipOutputStream(bos);

        try {
            zos.putNextEntry(new ZipEntry("file-b.txt"));
            byte[] b =  new byte[100000000];
            for (int i=0; i <= 60; ++i) {
                zos.write(b);
            }
            zos.closeEntry();
        }
        finally {
            zos.close();
        }
        System.out.println("Done");
    }
}
```
So far it looks like Java makes zip64 files (zip files for contents that are bigger than 2GiB) in a way that's not quite consistent with other zip files.

(Specifically, we assumed that zip64 member files have a local header with some magic values that say the compressed and uncompressed sizes of the files are equal to each other and equal to `0xffffffff`, but for Java, the sizes seem to be both 0)

**Comment by vitkovay** [comment]
In my zip archive, I can fix this by sending "True" to "get_crc_32_expected_from_data_descriptor" for a large file, but this is a one-time solution.

**Comment by vitkovay** [comment]
I think to fix error need to count the distance between b'PK\x07\x08' and b'PK\x03\x04' for each file and send it in "get_crc_32_expected_from_data_descriptor".

**Comment by michalc** [comment]
> In my zip archive, I can fix this by sending "True" to "get_crc_32_expected_from_data_descriptor" for a large file, but this is a one-time solution.

Ah this is helpful... but the tricky thing is is that I'm realising that I'm not sure how to do this in a robust way dynamically. From what I can tell so far, nothing in the stream before the data identifies the stream as zip64. Will ponder more...

**Comment by michalc** [comment]
Although there is more to do, have released a new version with the changes in https://github.com/uktrade/stream-unzip/pull/34 that I think improves the situation drastically with ZIPs created by Java.

It's not perfect, there is still more to be done - for example ZIPs created by Java that have a member file of exactly 4294967295 bytes in length won't yet work.

**Comment by michalc** [comment]
Ah - I think found a way to deal with member files that are exactly 4294967295 bytes long https://github.com/uktrade/stream-unzip/pull/35, and released in v0.0.72

Let me know if it now works for you?

**Comment by vitkovay** [comment]
Yes, it perfectly works for me. Thanks!
